mobile/library/jni/jni_utility.cc

```diff
@@ -426,5 +426,103 @@ absl::flat_hash_map<std::string, std::string> javaMapToCppMap(JniHelper& jni_hel
+        jni_helper.callObjectMethod(java_map_object.get(), java_map_get_method_id, java_key.get());
+    if (existing_value == nullptr) { // the key does not exist
+      // Create a new list.
+      auto java_list_object = jni_helper.newObject(java_list_class.get(), java_list_init_method_id);
+      jni_helper.callBooleanMethod(java_list_object.get(), java_list_add_method_id,
+                                   java_value.get());
+      // Put the new list into the map.
+      auto ignored = jni_helper.callObjectMethod(java_map_object.get(), java_map_put_method_id,
+                                                 java_key.get(), java_list_object.get());
+    } else {
+      // Update the existing list.
+      jni_helper.callBooleanMethod(existing_value.get(), java_list_add_method_id, java_value.get());
+    }
+
+    return Http::HeaderMap::Iterate::Continue;
+  });
+
+  return java_map_object;
+}
+
+void javaHeadersToCppHeaders(JniHelper& jni_helper, jobject java_headers,
+                             Http::RequestOrResponseHeaderMap& cpp_headers) {
+  auto java_map_class = jni_helper.getObjectClass(java_headers);
+  auto java_entry_set_method_id =
+      jni_helper.getMethodId(java_map_class.get(), "entrySet", "()Ljava/util/Set;");
+  auto java_entry_set_object = jni_helper.callObjectMethod(java_headers, java_entry_set_method_id);
+
+  auto java_set_class = jni_helper.getObjectClass(java_entry_set_object.get());
+  auto java_map_entry_class = jni_helper.findClass("java/util/Map$Entry");
+
+  auto java_map_iter_method_id =
+      jni_helper.getMethodId(java_set_class.get(), "iterator", "()Ljava/util/Iterator;");
+  auto java_map_get_key_method_id =
+      jni_helper.getMethodId(java_map_entry_class.get(), "getKey", "()Ljava/lang/Object;");
+  auto java_map_get_value_method_id =
+      jni_helper.getMethodId(java_map_entry_class.get(), "getValue", "()Ljava/lang/Object;");
+
+  auto java_iter_object =
+      jni_helper.callObjectMethod(java_entry_set_object.get(), java_map_iter_method_id);
+  auto java_iterator_class = jni_helper.getObjectClass(java_iter_object.get());
+  auto java_iter_has_next_method_id =
+      jni_helper.getMethodId(java_iterator_class.get(), "hasNext", "()Z");
+  auto java_iter_next_method_id =
+      jni_helper.getMethodId(java_iterator_class.get(), "next", "()Ljava/lang/Object;");
+
+  auto java_list_class = jni_helper.findClass("java/util/List");
+  auto java_list_size_method_id = jni_helper.getMethodId(java_list_class.get(), "size", "()I");
+  auto java_list_get_method_id =
+      jni_helper.getMethodId(java_list_class.get(), "get", "(I)Ljava/lang/Object;");
+
+  while (jni_helper.callBooleanMethod(java_iter_object.get(), java_iter_has_next_method_id)) {
+    auto java_entry_object =
+        jni_helper.callObjectMethod(java_iter_object.get(), java_iter_next_method_id);
+    auto java_key_object =
+        jni_helper.callObjectMethod<jstring>(java_entry_object.get(), java_map_get_key_method_id);
```

Comment:
I think at this point you're more familiar with JNI than I am so any perf concerns here?  It looks like it may have more JNI calls than the legacy javaArrayOfObjectArrayToEnvoyMap and I thought they were more expensive but I assume a factor of 2x isn't problematic compared to the bulk of the data copies right?

Reply:
Yeah having more JNI calls is usually not the issue compared to having to extra data copies.